compiler/rustc_const_eval/src/util/caller_location.rs

```diff
@@ -20,7 +20,10 @@ fn alloc_caller_location<'tcx>(
     // This can fail if rustc runs out of memory right here. Trying to emit an error would be
     // pointless, since that would require allocating more memory than these short strings.
     let file = if loc_details.file {
-        ecx.allocate_str(filename.as_str(), MemoryKind::CallerLocation, Mutability::Not).unwrap()
+        let mut name = filename.as_str().to_string();
+        name.push('\0');
+
+        ecx.allocate_str(&name, MemoryKind::CallerLocation, Mutability::Not).unwrap()
```

Comment:
Would be nice to mark the string as cstring in the compiled object files to allow the linker to deduplicate it.

Reply:
Is that done by using a different `MemoryKind`?

Reply:
I don't think we currently have a way to do this, but it would make sense to me to do it by adding a new `MemoryKind` and then special casing this in the codegen backends.

Reply:
I don't want to complicate this PR too much. That sounds like it could happen as a follow-up.